core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,555 @@
+      }
+      LockPlaceHolder wantToRelease = lockOwners.get(canceled.lockOwner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", canceled.lockOwner);
+         }
+         //nothing to release
+         return;
+      }
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", canceled.lockOwner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   private boolean casAcquire(LockPlaceHolder lockPlaceHolder) {
+      return fieldUpdater.compareAndSet(this, null, lockPlaceHolder);
+   }
+
+   private boolean casRelease(LockPlaceHolder lockPlaceHolder) {
+      return fieldUpdater.compareAndSet(this, lockPlaceHolder, null);
+   }
+
+   private void remove(Object lockOwner) {
+      lockOwners.remove(lockOwner);
+   }
+
+   private void triggerReleased() {
+      if (releaseRunnable != null) {
+         releaseRunnable.run();
+      }
+   }
+
+   private void tryAcquire() {
+      do {
+         LockPlaceHolder nextPending = pendingRequest.peek();
+         if (trace) {
+            log.tracef("Try acquire. Next in queue=%s. Current=%s", nextPending, current);
+         }
+         if (nextPending == null) {
+            return;
+         }
+         if (casAcquire(nextPending)) {
+            //we set the current lock owner, so we must remove it from the queue
+            pendingRequest.remove(nextPending);
+            if (nextPending.setAvailable()) {
+               if (trace) {
+                  log.tracef("%s successfully acquired the lock.", nextPending);
+               }
+               //successfully acquired
+               deadlockCheck();
+               return;
+            }
+            if (trace) {
+               log.tracef("%s failed to acquire (invalid state). Retrying.", nextPending);
+            }
+            //oh oh, probably the nextPending Timed-Out. we are going to retry with the next in queue
+            casRelease(nextPending);
+         } else {
+            if (trace) {
+               log.tracef("Unable to acquire. Lock is held.");
+            }
+            //other thread already set the current lock owner
+            return;
+         }
+      } while (true);
+   }
+
+   private LockPlaceHolder createLockInfo(Object lockOwner, long time, TimeUnit timeUnit) {
+      return new LockPlaceHolder(lockOwner, timeService.expectedEndTime(time, timeUnit));
+   }
+
+   private class LockPlaceHolder implements ExtendedLockPromise {
+
+      private final AtomicReferenceFieldUpdater<LockPlaceHolder, LockState> stateUpdater;
+
+      private final Object lockOwner;
+      private final long timeout;
+      private final AtomicBoolean cleanup;
```

Comment:
I am not a big fan of that for something that is really just a boolean.  However I wonder if this can be done differently.  It seems this boolean is only to protect running the release runnable.  Is it instead possible to have the remove return the previous value or a boolean and then optionally run the runnable only if it removed something?  Then you can remove this boolean completely.  I haven't had a chance to read over all of the classes to see if that makes sense or not, let me know if it doesn't

Something like:
```
 private void cleanup() {
   if (remove(lockOwner) != null) {
      triggerReleased();
   }
}
```